File: app/src/main/java/net/pmellaaho/rxapp/ui/RepoInputFragment.java
package net.pmellaaho.rxapp.ui;

import android.app.Activity;
import android.os.Bundle;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;

import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;

import net.pmellaaho.rxapp.R;
import net.pmellaaho.rxapp.RxApp;
import net.pmellaaho.rxapp.databinding.FragmentRepoInputBinding;
import net.pmellaaho.rxapp.model.Contributor;
import net.pmellaaho.rxapp.model.ContributorsModel;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import timber.log.Timber;

import static com.google.common.base.Strings.isNullOrEmpty;

public class RepoInputFragment extends Fragment {
    public static final String OWNER = "square";
    private static final String REPO = "retrofit";
    private static final String REQUEST_PENDING = "requestPending";

    private OnRepoInputListener mListener;

    private Button mButton;
    private EditText mRepoEdit;
    private ProgressBar mProgress;
    private View mErrorText;

    private CompositeDisposable mDisposables = new CompositeDisposable();

    ContributorsModel mModel;
    private boolean mRequestPending;

    public interface OnRepoInputListener {
        void onDataReady(String repo);
    }

    public class MyHandler {
        public void clicked(View v) {
            mButton.setEnabled(false);
            fetchData();
        }

        public void afterTextChanged(Editable s) {
            String newRepo = s.toString();

            if (!isNullOrEmpty(newRepo)) {
                mButton.setEnabled(true);
            } else {
                mButton.setEnabled(false);
            }
        }
    }

    public RepoInputFragment() {
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Timber.d("unsubscribe");
        mDisposables.dispose();
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        FragmentRepoInputBinding binding = DataBindingUtil.inflate(inflater, R.layout
                .fragment_repo_input, container, false);

        binding.setHandler(new MyHandler());

        mButton = binding.startBtn;
        mRepoEdit = binding.repoEdit;
        if (savedInstanceState == null) {
            mRepoEdit.setText(REPO);
        }

        mProgress = binding.progress;
        mErrorText = binding.errorText;

        mModel = ((RxApp) getActivity().getApplication()).component().contributorsModel();
        return binding.getRoot();
    }

    private void fetchData() {
        Timber.d("Fetch data from RepoInputFragment");
        mRequestPending = true;
        mProgress.setVisibility(View.VISIBLE);
        mErrorText.setVisibility(View.INVISIBLE);

        mDisposables.add(
                mModel.getContributors(OWNER, mRepoEdit.getText().toString())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new ContributorsSubscriber()));
    }


    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(REQUEST_PENDING, mRequestPending);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState != null && savedInstanceState.getBoolean(REQUEST_PENDING, false)) {

            if (mModel.getRequest() != null) {
                Timber.d("Subscribe to pending request");
                mDisposables.add(
                        mModel.getRequest()
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribeWith(new ContributorsSubscriber()));
            }
        }
    }

    private class ContributorsSubscriber extends DisposableObserver<List<Contributor>> {

        @Override
        public void onNext(List<Contributor> contributors) {
            Timber.d("received data from model");
        }

        @Override
        public void onComplete() {
            Timber.d("request completed");
            mRequestPending = false;
            mProgress.setVisibility(View.INVISIBLE);

            if (mListener != null) {
                mListener.onDataReady(mRepoEdit.getText().toString());
            }
        }

        @Override
        public void onError(Throwable e) {
            Timber.e(e, "request failed");
            mProgress.setVisibility(View.INVISIBLE);
            mErrorText.setVisibility(View.VISIBLE);
            mButton.setEnabled(true);
            mModel.reset();
        }
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            mListener = (OnRepoInputListener) activity;
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString()
                    + " must implement OnRepoInputListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }
}

File: app/src/main/java/net/pmellaaho/rxapp/ui/RepoInputFragment.kt
package net.pmellaaho.rxapp.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.unit.dp
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.ui.AppBarConfiguration
import com.google.android.material.composethemeadapter.MdcTheme
import dagger.hilt.android.AndroidEntryPoint
import net.pmellaaho.rxapp.R
import net.pmellaaho.rxapp.databinding.FragmentRepoInputBinding
import net.pmellaaho.rxapp.ui.ContributorsViewModel.ViewState.*
import timber.log.Timber

const val OWNER = "square"
private const val REPO = "retrofit"

@AndroidEntryPoint
class RepoInputFragment : Fragment() {
    private val viewModel: ContributorsViewModel by activityViewModels()
    private lateinit var binding: FragmentRepoInputBinding

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_repo_input, container, false)

        binding.composeView.apply {
            // Dispose the Composition when the view's LifecycleOwner is destroyed
            setViewCompositionStrategy(
                ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed
            )

            setContent {
                // entering the Compose world!
                RepoInputFragmentScreen(viewModel)
            }
        }

        return binding.root
    }

    @Composable
    private fun RepoInputFragmentScreen(viewModel: ContributorsViewModel) {
        val state: ContributorsViewModel.ViewState by viewModel.state.observeAsState(initial = EnterRepo)

        var text by rememberSaveable { mutableStateOf("") }
        val onTextChanged: (String) -> Unit = { text = it }

        MdcTheme {
            Scaffold(
                topBar = {
                    TopAppBar(
                        title = { Text(text = "Contributors") },
                    )
                },
                content = {
                    when (state) {
                        is Loading -> ProgressIndicator()
                        is EnterRepo -> {
                            RepoInputFragmentContent(
                                modifier = Modifier.padding(12.dp),
                                text,
                                onTextChanged
                            )
                        }

                        is ShowList -> { /* NoOp */ }
                        is Error -> {
                            Timber.e("request failed")
                            ErrorMessage()
                        }
                    }
                }
            )
        }
    }

    @Composable
    private fun RepoInputFragmentContent(
        modifier: Modifier = Modifier,
        text: String,
        onTextChanged: (String) -> Unit
    ) {
        Column(
            modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())) {
            Text(
                style = MaterialTheme.typography.h5,
                text = "Show the contributors to open source projects by Square"
            )

            val extraPadding = Modifier.padding(32.dp)
            OutLinedTextField(
                modifier = extraPadding.align(Alignment.CenterHorizontally),
                text = text,
                onTextChanged = onTextChanged
            )

            Button(
                modifier = extraPadding.align(Alignment.End),
                enabled = text.isNotEmpty(), // todo: disables state looks bad
                onClick = { fetchData(text) }) {
                Text(
                    text = "Start"
                )
            }
        }
    }

    @Composable
    fun OutLinedTextField(
        modifier: Modifier,
        text: String,
        onTextChanged: (String) -> Unit
    ) {
        OutlinedTextField(
            modifier = modifier,
            value = text,
            label = { Text(text = "repository name") },
            onValueChange = onTextChanged
        )
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val navController = findNavController()
        val appBarConfiguration = AppBarConfiguration(navController.graph)

        viewModel.navigateToList.observe(viewLifecycleOwner) { event ->
            if (event.getContentIfNotHandled() == true) {
                findNavController().navigate(R.id.listFragment)
            }
        }
    }

    private fun fetchData(repoName: String) {
        Timber.d("Fetch data from RepoInputFragment")
        viewModel.fetchContributors(OWNER, repoName)
    }

}